### ISSUE TITLE
Require an explicit mode for weather integration and bucket into three modes

### ISSUE THREAD

**Issue by MatthewFlamm** [home-assistant/architecture#382]
## Context

Currently, the weather integration supports hourly and daily weather forecasts.  Currently, the differentiator is whether a low temperature in `templow` is supplied by the backend and or how often the forecast is supplied.  This is cumbersome for the frontend to determine what kind of display is needed based on the availability and structure of the data.  It can also leave to issues such as this [one](https://github.com/home-assistant/frontend/issues/5792) where the structure of the backend data wasn't expected despite being a once daily forecast.  The fix required checking multiple forecasts to infer the intent.

There are currently three broad forecast types.  One is a once daily with a low temperature, the other is 'hourly' which is the type for any other forecast besides daily.  There is a middle class of forecasts that are not served well by either these two types: twice daily forecasts, ie a separate forecast for day and a separate forecast for night.  This forecast is more accurately referred to by the date and whether the forecast is for the day or night, e.g. Sun and Sun Night.  But the current system treats it as hourly.

## Proposal

Require weather entities to supply a `mode` attribute that can have one of three values: `hourly`, `daily`, `daynight`.  `daily` forecasts can optionally supply a `templow` value for forecasts.  `daynight` forecasts will require a `is_daytime` boolean to differentiate.  `hourly` forecasts will work the same as the current state.

## Consequences

It is easier to display data from weather integrations and avoid issues when subtle differences in data structure and type arise.  It requires all weather integrations to add one more attribute and the frontend card to use the new attribute to display the data.

**Comment by zsarnett** [comment]
Working with the Weather Card recently this was a very big problem that can be easily solved by the backend supplying which forecast is being sent. Specifically for DarkSky you actually have to set it in your config. If that was sent as an attribute on the entity it would make creating an agnostic weather card much easier.

**Comment by zsarnett** [comment]
Minimal changes I believe should be supplying the mode. `hourly`, `daily`, and `daynight`.

**Comment by zsarnett** [comment]
https://github.com/home-assistant/frontend/pull/6065

Another issue this would resolve if the Backend Supplied the mode. Instead of fixing each one a different way using guessing in the frontend we can fix them all and make the frontend code less complicated with this approach.

**Comment by MatthewFlamm** [comment]
> Minimal changes I believe should be supplying the mode. `hourly`, `daily`, and `daynight`.


I think this suggestion just removes the required `is_daytime` boolean for `daynight` entities. 
How will the frontend know which is which?  It isn't clear to me.  The timestamps are confusing in this case as Monday night forecast predominately occurs in Tuesday.

**Comment by rajlaud** [comment]
I also recommend having an `is_daytime` boolean. In the NWS forecasts, for example, the first forecast is short term, rather than at a fixed time, followed by 6:00 a.m. and 6:00 p.m. forecasts. To correctly distinguish day time and night time forecasts using timestamps, 6:00 p.m. precisely needs to be the dividing line. Other forecasts might be different, and the frontend would have to know them all.

**Comment by MatthewFlamm** [comment]
nws is currently the only integration offering a day and night forecast, but environment_canada is discarding it's night forecast to force it to be a daily forecast [here](https://github.com/home-assistant/core/blob/da6a99b333dfe939f8d16f017b3ffa85fd30f52b/homeassistant/components/environment_canada/weather.py#L176-L204), so it could probably support this mode as well.  Pinging @michaeldavie to confirm whether this is true.

**Comment by michaeldavie** [comment]
Yes, it would actually be simpler for me to handle the overnight forecasts as equivalent to the daytime ones. My backend library already does this, and I have to do some contortions to shoehorn the data into the form Home Assistant expects.

**Comment by MatthewFlamm** [comment]
https://github.com/home-assistant/core/pull/38725 adds another integration with day+night forecasts.

**Comment by zsarnett** [comment]
I would like us to decide a path for this as this will help shape the front end code

**Comment by frenck** [comment]
>  But the NWS integration has just gone ahead and implemented it as if it was approved.

It is implemented as a local solution for an integration, not as a global architectural change. This proposal has not been approved or implemented at this point.

**Comment by Makin-Things** [comment]
> This proposal has not been approved or implemented at this point.
But the NWS integration which is in core has implemented it and it has driven a change though to the forecast card.

**Comment by frenck** [comment]
@Makin-Things There is no architectural global change made as proposed in this issue.

**Comment by Makin-Things** [comment]
Also I am not trying to be confrontational, but it took a long time to find out why the NWS integration was doing things differently.
The first question though is this the best place to be discussing this?

**Comment by frenck** [comment]
> Also I am not trying to be confrontational, but it took a long time to find out why the NWS integration was doing things differently.

That is possible, doesn't mean it standardized.

> The first question though is this the best place to be discussing this?

Your original question was about this architectural proposal, which has been answered, as its not applied.

If you have other questions, please use our community forums or Discord chat.

**Comment by emontnemery** [comment]
Suggested answers, as per the original proposal by @MatthewFlamm:
Require weather entities to supply a mode attribute that can have one of three values: hourly, daily, daynight:
- daily forecasts can optionally supply a templow value for forecasts
- daynight forecasts require an is_daytime boolean to differentiate between day and night
- hourly forecasts are unchanged

**Comment by emontnemery** [comment]
## Suggestion:

Instead of adding a mode on the entity level, add new attributes `forecast_hourly` + `forecast_daily` which the entity can fill depending on what's received from the underlying service.
This would then allow displaying the data like this (for example):
![image](https://user-images.githubusercontent.com/14281572/177850822-b94405ee-2679-43b8-be7c-c7bf2880dea2.png)
